### ISSUE TITLE
Proposal: cache the /list endpoint for a short period of time?

### ISSUE THREAD

**Issue by arschles** [gomods/athens#893]
**Is your feature request related to a problem? Please describe.**

The `/list` endpoint contacts the upstream VCS to get version numbers, so that corresponds to one `git` (or `hg` etc...) operation per dependency. For a public deployment of athens, I believe that's untenable.

**Describe the solution you'd like**

I think it'd be a good idea to add an option to the proxy to cache the `/list` endpoint. That would mean that it fetches the version list directly from storage until a TTL expires. At that point it would do the list request to the upstream once, refresh storage, and reset the TTL. If there are multiple proxy servers running in a cluster, they would use #760 to ensure only one refresh is done after the TTL.

If we did this, we would need to tell package authors that if they push a new commit or release a new tag, it won't show up in Athens until the TTL expires. We could let them trigger a refresh for their module with #104, but I'm not sure if that's a reasonable solution.

**Describe alternatives you've considered**

I haven't thought of anything else that would reduce upstream requests for each `/list` call.

**Additional context**

I'm researching and preparing for #772 and found that some way to tell whether a response was cached or not would be really helpful for us.

**Comment by robjloranger** [comment]
I think some level of caching would be a big bonus here. Could we have manual invalidation on new version pushes?

**Comment by arschles** [comment]
@robjloranger yea, I think we could definitely have that. Maybe some variation on the webhook that got proposed in #104 could do the invalidation - what do you think?

**Comment by robjloranger** [comment]
Exactly 😁 some endpoint to say some module was "updated" and we can even check to see if it does have a new version, and if so update the list cache.

**Comment by arschles** [comment]
Update: https://github.com/gomods/athens/issues/1068 is the endpoint that @robjloranger talks about in https://github.com/gomods/athens/issues/893#issuecomment-443395041

I am going to start prototyping this now

**Comment by arschles** [comment]
We've discussed this a few times and decided this caching logic shouldn't live inside of Athens. We are discussing an "offline mode" in #1506, which will treat the list endpoint as immutable. If Athens gets an offline mode, then caching will become unnecessary if you run it in that configuration. If you were to still run it in online mode configuration and still want caching, the best thing to do would be to use a caching proxy. There are several widely used open source ones and several cloud based ones. I have also built a "toy" one [here](https://github.com/arschles/lathens) for reference.

I am going to close this.
